api/src/main/java/org/pmiops/workbench/cohortbuilder/FieldSetQueryBuilder.java

```diff
@@ -0,0 +1,320 @@
+    if (columnFilter.getValue() != null || columnFilter.getValueNumber() != null
+        || columnFilter.getValueDate() != null || columnFilter.getValueNull() != null) {
+      throw new BadRequestException("Can't use in operator with single value filter");
+    }
+    List<BigDecimal> valueNumbers = columnFilter.getValueNumbers();
+    List<String> valueStrings = columnFilter.getValues();
+
+    if (valueNumbers == null || valueNumbers.isEmpty()) {
+      if (valueStrings == null || valueStrings.isEmpty()) {
+        throw new BadRequestException("Can't use in operator without values or valueNumbers");
+      } else {
+        if (!columnConfig.type.equals(ColumnType.STRING)) {
+          throw new BadRequestException("Can't use values with column " + columnConfig.name
+              + " of type " + columnConfig.type);
+        }
+        paramMap.put(paramName, QueryParameterValue.array(valueStrings.toArray(new String[0]),
+            String.class));
+      }
+    } else {
+      if (valueStrings == null || valueStrings.isEmpty()) {
+        if (!columnConfig.type.equals(ColumnType.INTEGER)) {
+          throw new BadRequestException("Can't use valueNumbers with column " + columnConfig.name
+              + " of type " + columnConfig.type);
+        }
+        paramMap.put(paramName, QueryParameterValue.array(
+            valueNumbers.stream().map(BigDecimal::longValue).collect(
+                Collectors.toList()).toArray(new Long[0]),
+            Long.class));
+      } else {
+        throw new BadRequestException("Can't use both values and valueNumbers");
+      }
+    }
+    sqlBuilder.append(columnFilter.getColumnName());
+    sqlBuilder.append(" in unnest(${");
+    sqlBuilder.append(paramName);
+    sqlBuilder.append("})\n");
+
+  }
+
+  private void handleColumnFilter(ColumnFilter columnFilter, TableConfig tableConfig,
+      StringBuilder sqlBuilder, Map<String, QueryParameterValue> paramMap) {
+    if (Strings.isNullOrEmpty(columnFilter.getColumnName())) {
+      throw new BadRequestException("Missing column name for column filter");
+    }
+    ColumnConfig columnConfig = findColumn(tableConfig, columnFilter.getColumnName());
+    if (columnConfig == null) {
+      throw new BadRequestException("Column " + columnFilter.getColumnName() + " does not exist");
+    }
+    Operator operator = columnFilter.getOperator();
+    if (operator == null) {
+      operator = Operator.EQUAL;
+    }
+
+    if (operator.equals(Operator.IN)) {
+      handleInClause(columnFilter, columnConfig, sqlBuilder, paramMap);
+    } else {
+      handleComparison(columnFilter, columnConfig, sqlBuilder, paramMap);
+    }
+  }
+
+  private void handleColumnFilters(List<ColumnFilter> columnFilters, TableConfig tableConfig,
+      StringBuilder sqlBuilder, Map<String, QueryParameterValue> paramMap) {
+    boolean first = false;
+    if (columnFilters.isEmpty()) {
+      throw new BadRequestException("Empty column filter list is invalid");
+    }
```

Comment:
Does it make sense to move this check higher? Does it make sense to validate the Request is valid before starting to build the query?

Reply:
Maybe. I wanted to avoid having to visit everything twice... and we still probably are going to want some kind of a check / assert here to make sure we don't generate invalid SQL.